After starting running, keep sending keep_alive until receive stop

MateusPedrosa opened the pull request:
Depends on:
- [ ] https://github.com/tidewise/drivers-genset_whisperpower_ddc/pull/4

- Change implementation to follow the prinple used in
drivers/genset_whisperpower_ddc/src/Main.cpp. If receive start command
on the input port, start the generator and then keep sending keep_alive
command every time a new valid frame is received.
- Stop checking the command on the received frame to consider it valid.
- Add respective tests.

MateusPedrosa committed:
After starting running, keep sending keep_alive until receive stop

-Change implementation to follow the prinple used in
drivers/genset_whisperpower_ddc/src/Main.cpp. If receive start command
on the input port, start the generator and then keep sending keep_alive
command every time a new valid frame is received.
-Stop checking the command on the received frame to consider it valid.
-Add respective tests.

doudou commented on tasks/Task.cpp:
```diff
@@ -60,8 +60,8 @@ void Task::updateHook()
 
 void Task::processIO()
 {
+    ControlCommand sentCommand = CONTROL_CMD_NO_COMMAND;
```
Declare local variables close to where they are to be used (i.e. just before the `if(m_running)`)

doudou commented on tasks/Task.hpp:
```diff
@@ -33,6 +33,8 @@ tasks/Task.cpp, and will be put in the genset_whisperpower_ddc namespace.
 
         std::unique_ptr<VariableSpeedMaster> m_driver;
 
+        bool m_running = false;
```
Because there is a manual start button, we should initialize m_running on start, or the software might shut down a running generator on reboot.

doudou commented on tasks/Task.cpp:
```diff
@@ -78,25 +78,34 @@ void Task::processIO()
      * don't send control command
      */
-    if (frame.targetID == variable_speed::TARGET_ADDRESS && frame.sourceID == variable_speed::SOURCE_ADDRESS) {
+    if (frame.targetID == variable_speed::PANELS_ADDRESS && frame.sourceID == variable_speed::DDC_CONTROLLER_ADDRESS) {
```
Avoid deeply nested ifs. Return early with

~~~
if (frame.targetID != ... || frame.sourceID != ...) {
    return;
}
~~~

doudou commented on tasks/Task.cpp:
```diff
@@ -78,25 +78,34 @@ void Task::processIO()
+        if (m_running) {
+            sentCommand = CONTROL_CMD_KEEP_ALIVE;
+        }
+        while (_control_cmd.read(m_control_cmd) == RTT::NewData) {
```
You don't need both m_control_cmd and m_running. Actually, I think you should just keep m_control_cmd (renamed to m_last_command).

We read in attributes when the read value needs to be reused, or in critical codepaths to avoid memory allocations. But the control command has no allocation so you could have defined it as a local variable. 

Now, moving this logic to a separate method and removing m_running would dramatically reduce the number of branches (a good metric for "simplify"). In particular, it avoids the sentCommand temporary, which is really hard to follow:

In `updateHook`:

~~~
m_last_cmd = processStartStopCommand();
~~~

The new method:

~~~
ControlCommand Task::processStartStopCommand() {
    ControlCommand cmd;
    if (_control_cmd.read(cmd) == RTT::NoData) {
        return m_last_cmd;
    }

    if (cmd != m_last_cmd) {
        m_driver->sendControlCommand(cmd);
        return cmd
    }

    if (m_last_cmd == CONTROL_CMD_START) {
        m_driver->sendControlCommand(CONTROL_CMD_KEEPALIVE);
    }
    return m_last_cmd;
}
~~~

doudou commented:
There is one last bit, that I missed in the first review, but is now very obvious:

Using `ControlCommand` as the type of the input port does not make sense anymore. It has 4 different values, two of which we don't process, and one of them (`_START`) is actually interpreted by the component to mean something else (_KEEPALIVE).

Just make it a boolean.

MateusPedrosa committed:
Simplify the logic for sending control command

- Make input port a boolean
- Remove m_control_cmd
- Add processStartStopCommand() method

doudou commented on tasks/Task.cpp:
```diff
@@ -50,6 +51,9 @@ bool Task::startHook()
     if (! TaskBase::startHook())
         return false;
+
+    m_running = false;
```
You are still assuming that the command on start is "stop". This is wrong, as if for whatever reason the software needs to reboot, it will shut down the generator (assume a faulty battery, what happens then ?).

The component needs to keep the state of the generator "as it found it on start".

What would be reasonable doing is, in startHook, to wait receiving the generator state and initialize m_running using that information.

MateusPedrosa committed:
Check if is running based on GeneratorStatus in startHook

MateusPedrosa committed:
Check run signal bit instead of generator_status to know if is running

MateusPedrosa committed:
Add timeout in startHook

doudou committed:
test: fix starting the task

doudou committed:
refactor isRunning for clarity

doudou committed:
display exception received in startHook

doudou committed:
vscode: add debug target

doudou committed:
test: fix starting the component

poll is called too often, the frames are not separated by the driver
(which is time-based)

We do need join_all_waiting_work(false) or the poll block is not
called repeatedly

MateusPedrosa committed:
Send keep_alive if genset is running and there's no new data on control_cmd

MateusPedrosa committed:
Fix tests and clarify code

Separate code in smaller methods

doudou commented on test/task_test.rb:
```diff
@@ -47,25 +39,25 @@ def iodrivers_base_prepare(model)
-                     sample.rotation_speed.round(5)
-        assert_equal (0.01 * ((0x03 << 8) | 0x02)).round(5),
-                     sample.start_battery_voltage.round(5)
+        assert_equal(
```
There is `assert_in_delta` for this. `assert_in_delta x, y, precision`. You usually can omit the precision (the default is good for a lot of common cases)

Every test harness / framework has special assertions for floating-point comparison.

MateusPedrosa committed:
Use assert_in_delta for floating-point comparisons

doudou commented on tasks/Task.cpp:
```diff
@@ -119,4 +134,63 @@ void Task::cleanupHook()
+{
+    bool cmd;
+    if (_control_cmd.read(cmd) == RTT::NoData) {
+        if (m_running) {
+            m_driver->sendControlCommand(CONTROL_CMD_KEEP_ALIVE);
+        }
```
`NoData` should really mean "do not do anything"

MateusPedrosa replied:
The `control_cmd` port receives true or false to respectively start or stop the generator. After getting it running, the component should repeatedly send keep_alive even if there isn't anything on this input port, otherwise we would go back to having to manually send the keep_alive command every time.

MateusPedrosa replied:
Maybe you prefer to continuously send true to keep it running?

doudou replied:
Ah ... misunderstanding here.

`NoData` means nothing has ever been written on the port (or .clear() was called). When you read a port for which there has been some data written, you get OldData (and the value is re-read in the argument)

doudou commented on test/task_test.rb:
```diff
@@ -15,6 +15,7 @@
                   .deployed_as("genset_whisperpower_ddc")
         )
         @task.properties.io_read_timeout = Time.at(2)
+        self.expect_execution_default_timeout = 30
```
Do we really need this ?

MateusPedrosa committed:
Remove reader attribute and specific assertion methods for sending command frames

-Instead of having specific methods for asserting that a frame has been
sent, use assert_driver_processes_frame after writing a new command to
port

MateusPedrosa committed:
Check if generator is running in processIO by parsing generator_state frame

Every time the component receives a PACKET_GENERATOR_STATE_AND_MODEL
frame it verifies if the generator is running by parsing the frame and
checking the run signal bit. It does not assume the state has changed
simply because a new command frame was sent.

MateusPedrosa committed:
Simplify tests and add tests for changing the state during execution

MateusPedrosa committed:
Stop writing the same command multiple times on a single test

doudou commented on tasks/Task.cpp:
```diff
@@ -59,6 +60,8 @@ bool Task::startHook()
     if (! TaskBase::startHook())
         return false;
 
+    std::cout << "RUNNING: " << endl;
+
```
Leftover debug

doudou commented on tasks/Task.cpp:
```diff
@@ -4,6 +4,7 @@
 #include <iodrivers_base/ConfigureGuard.hpp>
 #include <genset_whisperpower_ddc/VariableSpeed.hpp>
 #include <genset_whisperpower_ddc/ControlCommand.hpp>
+#include <fstream>
```
Also leftover because of the cout ? You only need iostream for cout

MateusPedrosa replied:
Sorry, should have paid more attention. It was actually for a workaround I did before you showed me the correct location of the log file.

MateusPedrosa committed:
Remove leftovers from debug

doudou commented:
Thanks @MateusPedrosa. I'm going to merge it myself, I'd like to have a chance to have it built for tomorrow morning.
